Python/Product/PythonTools/PythonTools/LanguageServerClient/FileWatcher/Listener.cs

```diff
@@ -114,14 +114,11 @@ class Listener : IDisposable {
                 return;
             }
 
-            // _rpc can be disposed and set to null at any time, since all the calls using it are asynchronous.
-            // If the rpc has already been set to null, we're done.
-            // Otherwise, store a local copy of the rpc to avoid having to check for null every time it's used.
-            // Then, when using it, catch any expected exceptions to avoid a crash.
-            if (_rpc is null) {
+            // _rpc can be disposed and/or set to null at any time, since all the calls using it are asynchronous.
+            // Check for null and disposed now, and every time it's used.
+            if (_rpc is null || _rpc.IsDisposed) {
```

Comment:
Since this code is entered via `async void`, it can potentially run concurrently (not just asynchronously) on a different thread relative to the one that does `Dispose()` and `=null`, so this check still has a race condition between left and right-hand side. This can be mitigated by copying it to a local before checking, or by using nullable bools:
```cs
if (_rpc?.IsDisposed != false)  // null or true
```
But also, looking at the rest of the class, it looks like it might be possible to just make `_rpc` readonly, and drop the null checks altogether?

Reply:
Ahh, good catch. I didn't think about the rpc being set to null before the second check on that conditional, but yeah that's definitely true.

`_rpc` gets set to null in the `_rpc_Disconnected` callback, so I don't think it can be readonly, just because of that. The reason I added the null checks back in was because that's the existing pattern in the PythonLanguageClient. But I agree, the code is much better to read the other way. I'll add the local back in.